src/app/src/main/java/com/example/bookmark/SignupActivity.java

```diff
@@ -0,0 +1,200 @@
+
+        // get relevant views
+        userNameEditText = findViewById(R.id.signup_username_edit_text);
+        firstNameEditText = findViewById(R.id.signup_firstname_edit_text);
+        lastNameEditText = findViewById(R.id.signup_lastname_edit_text);
+        emailAddressEditText = findViewById(R.id.signup_email_edit_text);
+        phoneNumberEditText = findViewById(R.id.signup_phonenumber_edit_text);
+        signUpButton = findViewById(R.id.signup_signup_button);
+
+        userNameLayout = findViewById(R.id.signup_username_textInputLayout);
+        firstNameLayout = findViewById(R.id.signup_firstname_textInputLayout);
+        lastNameLayout = findViewById(R.id.signup_lastname_textInputLayout);
+        emailAddressLayout = findViewById(R.id.signup_email_textInputLayout);
+        phoneNumberLayout = findViewById(R.id.signup_phonenumber_textInputLayout);
+
+        ArrayList<EditText> formTexts = new ArrayList<>();
+        formTexts.add(userNameEditText);
+        formTexts.add(firstNameEditText);
+        formTexts.add(lastNameEditText);
+
+        ArrayList<TextInputLayout> formLayouts = new ArrayList<>();
+        formLayouts.add(userNameLayout);
+        formLayouts.add(firstNameLayout);
+        formLayouts.add(lastNameLayout);
+
+         signUpButton.setOnClickListener(view -> {
+            boolean invalidField = false;
+            // double check all form fields filled
+            for (int i = 0; i < formTexts.size(); i++) {
+                if (!validateEditTextEmpty(formTexts.get(i), formLayouts.get(i))) {
+                    invalidField = true;
+                }
+            }
+            // check email and phone validation
+            if (!checkIfEditTextValidEmail(emailAddressEditText, emailAddressLayout)) {
+                invalidField = true;
+            }
+            if (!checkIfEditTextValidPhoneNumber(phoneNumberEditText, phoneNumberLayout)) {
+                invalidField = true;
+            }
+            if (invalidField) {
+                return;
+            }
+
+            // check username is not in db already
+            String username = userNameEditText.getText().toString();
+            firebaseProvider.retrieveUserByUsername(username,
+                user -> {
+                    if (user == null) {
+                        // Get values from fields and create user object
+                        String firstName = firstNameEditText.getText().toString();
+                        String lastName = lastNameEditText.getText().toString();
+                        String emailAddress = emailAddressEditText.getText().toString();
+                        String phoneNumber = phoneNumberEditText.getText().toString();
+
+                        // add user and go back to main activity
+                        User newUser = new User(username, firstName, lastName, emailAddress, phoneNumber);
+                        addUser(newUser, firebaseProvider);
```

Comment:
Be careful doing this. `FirebaseProvider.storeUser` is an asynchronous method and it may not have finished executing before you navigate away from SignupActivity. You should move the code that creates an Intent and starts the MainActivity into the onSuccessListener.

As written, the following scenario could occur:
1. A user clicks sign up.
2. This activity starts the MainActivity.
3. The user tries logging in.
4. The login fails.
5. The user's account gets created or an error occurs.